<a href="https://colab.research.google.com/github/MRazin172/NLPSentimentAnalysis/blob/main/NLPSentimentAnalysis.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [17]:
import pandas as pd
import nltk
from nltk.tokenize import word_tokenize
from nltk.corpus import stopwords
from nltk.stem import WordNetLemmatizer
from sklearn.model_selection import train_test_split
from sklearn.feature_extraction.text import CountVectorizer
from sklearn.naive_bayes import MultinomialNB
from sklearn.metrics import accuracy_score, classification_report


In [18]:
nltk.download('punkt')
nltk.download('stopwords')
nltk.download('wordnet')

[nltk_data] Downloading package punkt to /root/nltk_data...
[nltk_data]   Package punkt is already up-to-date!
[nltk_data] Downloading package stopwords to /root/nltk_data...
[nltk_data]   Package stopwords is already up-to-date!
[nltk_data] Downloading package wordnet to /root/nltk_data...
[nltk_data]   Package wordnet is already up-to-date!


True

In [19]:
data = pd.read_csv('/content/Reddit_Data.csv')

In [20]:
data.head()

clean_comment  category
0   family mormon have never tried explain them t...         1
1  buddhism has very much lot compatible with chr...         1
2  seriously don say thing first all they won get...        -1
3  what you have learned yours and only yours wha...         0
4  for your own benefit you may want read living ...         1

the dataset has two columns, the first column has the Comments and the Second one indicates its Sentimental Label

In [21]:
data.shape

(37249, 2)

the Dataset has around 37K Comments along with its Sentimental Label

In [22]:
data.columns

Index(['clean_comment', 'category'], dtype='object')

1. 0 Indicating it is a Neutral Comment
2. 1 Indicating a Postive Sentiment
3. -1 Indicating a Negative Comment

In [23]:
data.isnull()

clean_comment  category
0              False     False
1              False     False
2              False     False
3              False     False
4              False     False
...              ...       ...
37244          False     False
37245          False     False
37246          False     False
37247          False     False
37248          False     False

[37249 rows x 2 columns]

In [24]:
data.isnull().sum()

clean_comment    100
category           0
dtype: int64

preprocessing the dataset

In [28]:
stop_words = set(stopwords.words('english'))
lemmatizer = WordNetLemmatizer()

def preprocess_text(text):
    if isinstance(text, str):
        tokens = word_tokenize(text.lower())
        tokens = [token for token in tokens if token.isalpha()]
        tokens = [token for token in tokens if token not in stop_words]
        tokens = [lemmatizer.lemmatize(token) for token in tokens]
    else:
        return ''

data['newclean_comment'] = data['clean_comment'].apply(preprocess_text)

we are first tokeizing the comments column and making it to lowre case.
then removing the non alphabetic tokens like punctuations.
also removing the stopwords. and
lemmatizing them


The preprocessed comment column is renamed as newclean_comment

In [27]:

X_train, X_test, y_train, y_test = train_test_split(data['newclean_comment'], data['category'], test_size=0.2, random_state=42)

vectorizer = CountVectorizer()
X_train_counts = vectorizer.fit_transform(X_train)
X_test_counts = vectorizer.transform(X_test)

clf = MultinomialNB()
clf.fit(X_train_counts, y_train)

y_pred = clf.predict(X_test_counts)
print("Accuracy:", accuracy_score(y_test, y_pred))
print(classification_report(y_test, y_pred))


Accuracy: 0.6140939597315436
              precision    recall  f1-score   support

          -1       0.67      0.42      0.52      1667
           0       0.86      0.37      0.52      2615
           1       0.55      0.92      0.69      3168

    accuracy                           0.61      7450
   macro avg       0.69      0.57      0.57      7450
weighted avg       0.69      0.61      0.59      7450



#inferences

The overall accuracy of the sentiment analysis model is approximately 61.4%.

for -1 which are negative comments the  precision is 67%, recall is 42%, and F1-score is 52%

for 0 which are neutrl commets the precision is 86%, recall is 37%, and F1-score is 52 percnt

for 1 which r positive the precision is 55%, recall is 92%, and F1-score is 69

overall the accuracy is only 61.4